src/core/lib/utils/blockAllocator/xallocator.cpp

```diff
@@ -149,219 +149,156 @@ static inline void lock_get() {
 /// ~XallocInitDestroy destructor calls this function automatically.
-extern "C" void xalloc_destroy() {
+void xalloc_destroy() {
     lock_get();
     std::unique_lock<std::mutex> lock(xalloc_mutex);
     {
+        for (auto& [k, a] : _allocators) {
 #ifdef STATIC_POOLS
-        for (usint i = 0; i < MAX_ALLOCATORS; i++) {
-            _allocators[i]->~Allocator();
-            _allocators[i] = 0;
-        }
+            a->~Allocator();
 #else
-        for (usint i = 0; i < MAX_ALLOCATORS; i++) {
-            if (_allocators[i] == 0)
-                break;
-            delete _allocators[i];
-            _allocators[i] = 0;
-        }
+            delete a;
 #endif
+            a = nullptr;
+        }
+        _allocators.clear();
     }
     lock_release();
-
     lock_destroy();
 }
 
 /// Get an Allocator instance based upon the client's requested block size.
 /// If a Allocator instance is not currently available to handle the size,
 ///  then a new Allocator instance is create.
 ///  @param[in] size - the client's requested block size.
-///  @return An Allocator instance that handles blocks of the requested
-///  size.
-extern "C" Allocator* xallocator_get_allocator(size_t size) {
-    // Based on the size, find the next higher powers of two value.
-    // Add sizeof(Allocator*) to the requested block size to hold the size
-    // within the block memory region. Most blocks are powers of two,
-    // however some common allocator block sizes can be explicitly defined
-    // to minimize wasted storage. This offers application specific tuning.
-    OPENFHE_DEBUG_FLAG(false);
+///  @return An Allocator instance that handles blocks of the requested size.
+Allocator* xallocator_get_allocator(size_t size) {
     size_t blockSize = size + sizeof(Allocator*);
-    if (blockSize > 256 && blockSize <= 396)
-        blockSize = 396;
-    else if (blockSize > 512 && blockSize <= 768)
-        blockSize = 768;
-    else
-        blockSize = nexthigher<size_t>(blockSize);
-    OPENFHE_DEBUG("finding allocator for blockSize " << blockSize);
-    Allocator* allocator = find_allocator(blockSize);
-
-#ifdef STATIC_POOLS
-
-    if (allocator == nullptr) {
-        std::cerr << "static pool allocator for blockSize " << blockSize << " not found! " << std::endl;
-    }
-    assert(allocator != nullptr);
-#else
-    // If there is not an allocator already created to handle this block size
-    if (allocator == nullptr) {
-        // Create a new allocator to handle blocks of the size required
-        allocator = new Allocator(blockSize, 0, 0, "xallocator");
-
-        // Insert allocator into array
-        insert_allocator(allocator);
-    }
-#endif
-
-    return allocator;
+    auto it = _allocators.lower_bound(blockSize);
+    if (it != _allocators.end())
+        return it->second;
+    OPENFHE_THROW(lbcrypto::config_error, std::string(__func__) + ": Exceeded max block size.");
```

Comment:
Use the new OPENFHE_THROW that doesn't need "lbcrypto::config_error, std::string(func)". Also ": Exceeded max block size." should be replaced with a more meaningful message